kafka_consumer_freshness_tracker/src/main/java/com/tesla/data/consumer/freshness/ConsumerFreshness.java

```diff
@@ -169,95 +170,151 @@ void run() {
-        if(partitions.size() > 0 && !anyEndOffsetFound){
-          LOG.warn("Skipping {}: {} because no end-offsets found for any of the {} topic/partitions",
-              client.getCluster(), consumerGroup, partitions.size());
-        }
+        completedConsumers.add(measureConsumer(client, workers, consumerGroup));
       }
-      metrics.lastClusterRunSuccessfulAttempt.labels(client.getCluster()).setToCurrentTime();
+    } catch (InterruptedException e) {
+      LOG.error("Interrupted while measuring consumers for {}", client.getCluster(), e);
+      return Futures.immediateFailedFuture(e);
     } catch (RejectedExecutionException e) {
       LOG.error("Failed to run freshness lookup for cluster {}, shutting down...", client.getCluster(), e);
       throw e;
-    } catch (IOException | InterruptedException | RuntimeException e) {
-      LOG.error("Failed to handle cluster {}", client.getCluster(), e);
-      metrics.burrowClustersConsumersReadFailed.labels(client.getCluster()).inc();
     }
-    return completed;
+
+    // if all the consumer measurements succeed, then we return the cluster name
+    // otherwise, Future.get will throw an exception representing the failure to measure a consumer (and thus the
+    // failure to successfully monitor the cluster).
+    return Futures.whenAllSucceed(completedConsumers).call(client::getCluster, this.executor);
+  }
+
+  /**
+   * Measure the freshness for all the topic/partitions currently consumed by the given consumer group. To maintain
+   * the existing contract, a consumer measurement fails ({@link Future#get()} throws an exception) only if:
+   *  - burrow group status lookup fails
+   *  - execution is interrupted
+   * Failure to actually measure the consumer is swallowed into a log message & metric update; obviously, this is less
+   * than ideal for many cases, but it will be addressed later.
+   *
+   * @param burrow access to burrow
+   * @param workers pool of workers available for querying Kafka
+   * @param consumerGroup name of the consumer group to measure
+   * @return a future representing the measurement of each topic-partition the consumer reads. If any partition
+   * failed to be measured the entire future (i.e. calls to {@link Future#get()}) will be considered a failure.
+   * @throws InterruptedException if the application is interrupted while waiting for an available worker
+   */
+  private ListenableFuture<List<Object>> measureConsumer(Burrow.ClusterClient burrow,
+                                                         ArrayBlockingQueue<KafkaConsumer> workers,
+                                                         String consumerGroup) throws InterruptedException {
+    Map<String, Object> status;
+    try {
+      status = burrow.getConsumerGroupStatus(consumerGroup);
+      Preconditions.checkState(status.get("partitions") != null,
+          "Burrow response is missing partitions, got {}", status);
+    } catch (IOException | IllegalStateException e) {
```

Comment:
do you still need to catch IllegalState?

Reply:
I'm hesitant to get rid of it right now - theoretically you would be be getting back an error (ie. non 200) response so you wouldn't run into this, but then again we have seen some weird stuff in practice with Burrow.

Not catching the checkState would cause the entire application to fail, rather than just the cluster or the consumer, so if the checkState stays, so should the catch for the exception.